src/coreclr/nativeaot/docs/compiling.md

```diff
@@ -153,3 +153,7 @@ Alpine
 ```sh
 apk add numactl-dev
 ```
+
+## NativeAOT on NixOS
+
+NativeAOT relies on hard-coded paths, such as `/lib64/ld-linux-x86-64.so.2`, which do not exist on NixOS. [Here](https://nixos.wiki/wiki/DotNET#NativeAOT) is an instruction for NixOS.
```

Comment:
@jkotas, maybe this is better:
```suggestion
For NixOS, follow the instructions at https://nixos.wiki/wiki/DotNET#NativeAOT.
```

which mentions `[nix-ld](https://github.com/Mic92/nix-ld) is needed` among other things.

Reply:
nix-ld is what allows to workaround that /lib64/ld-linux-x86-64.so.2 pass missing FWIW

Reply:
Dynamic linker is a requirement, so `ld` on any linux is needed to run NativeAOT published dynamic executable. If you are fulfilling the requirement as the doc suggests, it is not a workaround.

On Alpine aarch64 the interpreter is `/lib/ld-musl-aarch64.so.1`, on Ubuntu amd64 it is `/lib64/ld-linux-x86-64.so.2` and so on. I am not sure where the "hard-coded path" is coming from. Did you see in the output of some tools like `file` or `ldd`? You can check with readelf to be absolutely sure:

```sh
# run all these on NixOS
$ dotnet new console -n exe1
$ dotnet publish exe1 -o exe1/dist -p:PublishAot=true
$ readelf -l exe1/dist/exe1 | grep -A 2 INTERP
```

this is what I see on Alpine aarch64:

```
  INTERP         0x0000000000000270 0x0000000000400270 0x0000000000400270
                 0x000000000000001a 0x000000000000001a  R      0x1
      [Requesting program interpreter: /lib/ld-musl-aarch64.so.1]
```

Reply:
I invoked `ilc` with `lddtree` and it showed that particular path as a dependency. 

> If you are fulfilling the requirement as the doc suggests, it is not a workaround.

It is a workaround in a sense, that that path is not what we normally do on NixOS (and it was available under glibc path).

Reply:
What do you normally do on NixOS? What would it take to make native AOT to do what you normally do on NixOS?

Reply:
I'm not sure, to be honest. The dotnet SDK itself is packaged correctly and patches the linker loader path to glibc one, but external tools, like `ilc`, which are fetched by dotnet SDK and not NixOS, still need that path. Currently it is solved by NixOS users by symlinking the actual path to `/lib64/...` (and this isn't a NativeAOT-specific problem). 

The true way for us (NixOS users) would be replacing nuget with our wrapper of the tools, to provide them the correct environment with expected paths.

I'm not 100% sure if NativeAOT can trivially solve it, because this is likely the problem of any native tools fetched through nuget.

Reply:
> I invoked `ilc` with `lddtree`

Ah, so you are referring to `ilc`, `crossgen2` etc. all the pre-built binaries under `~/.nuget` directory, and not the binaries published by NativeAOT _on_ NixOS. The pre-built binaries (`ilc` etc.) in `runtime.linux-x64.microsoft.dotnet.ilcompiler` package were compiled on CentOS.

> What do you normally do on NixOS?

This article explains why pre-built binaries from other distro do not work on NixOS without `nix-ld`: https://blog.thalheim.io/2022/12/31/nix-ld-a-clean-solution-for-issues-with-pre-compiled-executables-on-nixos/. It is a typical way to run pre-built  binaries on NixOS, it is not a special workaround.

Reply:
> Ah, so you are referring to ilc, crossgen2 etc. all the pre-built binaries under ~/.nuget directory, and not the binaries published by NativeAOT on NixOS

Yes, sorry for possible confusion.

> It is a typical way to run pre-built binaries on NixOS, it is not a special workaround.

If .net tools were packed by someone for .net, it wouldn't be needed (and it isn't an issue with most programs). But like I said - it isn't a NativeAOT issue either. It's just happens that nobody (yet) made a convenient way to install dotnet tools without `dotnet tool install` (which is mutable and fetches non-nix-native executables, so to say). 

NativeAOT also relies on a few other deps, but they can be found from LD_LIBRARY_PATH (as I described in the link).

Reply:
> If .net tools were packed by someone for .net, it wouldn't be needed (and it isn't an issue with most programs). 

Yes, I think the portable way for custom tools like `ilc` (which were installed without `dotnet tool install`) is by creating an MSIL package (with `ilc.dll` and dependencies). Then change invocation in targets: https://github.com/dotnet/runtime/blob/47071da67320985a10f4b70f50f894ab411f4994/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L262 as `Command="&quot$(DOTNET_HOST_PATH)&quot; &quot;$(IlcToolsPath)\ilc.dll&quot ....`

Since this is NixOS-only issue (so far), it might not warrant creating a separate official package with MSIL variant.

The "official" NuGet way for organization/distros to deal with this kind of stuff is that you create nuget feed for your org, add source mapping (https://devblogs.microsoft.com/nuget/introducing-package-source-mapping/#package-source-mapping-example-configuration) in global `~/.nuget/NuGet/NuGet.Config` special-casing `runtime.linux-x64.microsoft.dotnet.ilcompiler` etc. and publish it with `same version` built on your org server to org nuget feed. This way, the consumer running `dotnet publish -p:PublishAot=true` on NixOS will not pull this package from nuget.org, but your org's server.